src/NUnitEngine/nunit.engine/Runners/DirectTestRunner.cs

```diff
@@ -86,10 +98,19 @@ protected override TestEngineResult LoadPackage()
             foreach (var subPackage in packages)
             {
                 var testFile = subPackage.FullName;
+
+                if (_assemblyResolver != null && !TestDomain.IsDefaultAppDomain()
+                    && subPackage.GetSetting(PackageSettings.ImageRequiresDefaultAppDomainAssemblyResolver, false))
+                {
+                    _assemblyResolver.AddPathFromFile(testFile);
+                }
+
                 IFrameworkDriver driver = driverService.GetDriver(TestDomain, testFile);
                 driver.ID = TestPackage.ID;
                 result.Add(driver.Load(testFile, subPackage.Settings));
                 _drivers.Add(driver);
+
+                _packageIndex.Add(driver, subPackage);
```

Comment:
It makes it a little more complicated. Each user of the resolver would have to remove any entry that they had created at the appropriate time. However, for extensions, the appropriate time probably coincides with process exit. Leaving entries shouldn't break anything, unless there are duplicate assembly names, but it would be inefficient.

I just merged my PR with master. I think you should update your branch. I don't think there will be conflicts and the two resolvers should work OK together.

Once your work is done, we can look at cleaning up the architecture. Since we are pushing for 3.2 next week, that would be afterwards.